lib/remap-lang/src/expression/arithmetic.rs

```diff
@@ -46,16 +43,28 @@ impl Expression for Arithmetic {
         use value::Kind;
         use Operator::*;
 
-        let kind = match self.op {
-            Or => self.lhs.type_def(state).kind | self.rhs.type_def(state).kind,
-            Multiply | Add => Kind::Bytes | Kind::Integer | Kind::Float,
-            Remainder | Subtract | Divide => Kind::Integer | Kind::Float,
-            And | Equal | NotEqual | Greater | GreaterOrEqual | Less | LessOrEqual => Kind::Boolean,
-        };
+        let lhs_def = self.lhs.type_def(state);
+        let rhs_def = self.rhs.type_def(state);
+        let type_def = lhs_def | rhs_def;
 
-        TypeDef {
-            fallible: true,
-            kind,
+        match self.op {
+            Or if lhs_def.kind.is_null() => rhs_def,
+            Or if !lhs_def.kind.is_null() && !lhs_def.kind.is_boolean() => lhs_def,
+            Or => type_def,
+            And if lhs_def.kind.is_null() => lhs_def.with_constraint(Kind::Boolean),
+            And => type_def
+                .fallible_unless(Kind::Null | Kind::Boolean)
+                .with_constraint(Kind::Boolean),
+            Equal | NotEqual => type_def.with_constraint(Kind::Boolean),
+            Greater | GreaterOrEqual | Less | LessOrEqual => type_def
+                .fallible_unless(Kind::Integer | Kind::Float)
+                .with_constraint(Kind::Boolean),
+            Subtract | Divide | Remainder => type_def
+                .fallible_unless(Kind::Integer | Kind::Float)
+                .with_constraint(Kind::Integer | Kind::Float),
+            Multiply | Add => type_def
+                .fallible_unless(Kind::Bytes | Kind::Integer | Kind::Float)
+                .with_constraint(Kind::Bytes | Kind::Integer | Kind::Float),
```

Comment:
I don't think this is quite what I was suggesting. `with_constraint` will override all the constraints of `type_def`. But, if we know that `type_def` is `Kind::Bytes`, then it should stay as `Kind::Bytes` and not have `Kind::Integer | Kind::Float` added on.

So 

`"some text" + "more text"` should constrain to just `Kind::Bytes`.
`.foo + "some text"` should be `Kind::Bytes`.
`3 + .foo` should be `Kind::Integer`
`.foo + .bar` should be `Kind::Bytes | Kind::Integer | Kind::Float`  because we have no further info on what each side should be.

I don't think it is a disaster with the current way, but I don't think there is much of a cost to being more precise here.

Reply:
I agree. It would be nice to restrict the type when we can.

Reply:
Good point, I'll make the change.

Reply:
I decided against changing this in this PR — given that the changes in this PR already move us in the right direction, and this PR has been blocking other pending work — but I agree we should make further improvements in a follow-up PR.

I filed #5246 as a tracking issue.